haystack/preview/components/fetchers/link_content.py

```diff
@@ -0,0 +1,163 @@
+
+        :param raise_on_failure: A boolean indicating whether to raise an exception when a failure occurs
+            during content extraction. If False, the error is simply logged and the program continues.
+            Defaults to False.
+        :param user_agents: A list of user agents to use when fetching content. Defaults to None, in which case a
+            default user agent is used.
+        :param retry_attempts: The number of times to retry fetching content. Defaults to 2.
+        :param timeout: The timeout in seconds for the request. Defaults to 3.
+        """
+        self.raise_on_failure = raise_on_failure
+        self.user_agents = user_agents or [DEFAULT_USER_AGENT]
+        self.current_user_agent_idx: int = 0
+        self.retry_attempts = retry_attempts
+        self.timeout = timeout
+
+        # register default content handlers that extract data from the response
+        self.handlers: Dict[str, Callable[[Response], Dict[str, Any]]] = defaultdict(lambda: text_content_handler)
+        self.handlers["text/html"] = text_content_handler
+        self.handlers["text/plain"] = text_content_handler
+        self.handlers["application/pdf"] = binary_content_handler
+        self.handlers["application/octet-stream"] = binary_content_handler
+
+        @retry(
+            reraise=True,
+            stop=stop_after_attempt(self.retry_attempts),
+            wait=wait_exponential(multiplier=1, min=2, max=10),
+            retry=(retry_if_exception_type((HTTPError, requests.RequestException))),
+            # This method is invoked only after failed requests (exception raised)
+            after=self._switch_user_agent,
+        )
+        def get_response(url):
+            # we need to copy because we modify the headers
+            headers = REQUEST_HEADERS.copy()
+            headers["User-Agent"] = self.user_agents[self.current_user_agent_idx]
+            response = requests.get(url, headers=headers, timeout=timeout or 3)
+            response.raise_for_status()
+            return response
+
+        self._get_response: Callable = get_response
+
+    def to_dict(self) -> Dict[str, Any]:
+        """
+        Serialize this component to a dictionary.
+        """
+        return default_to_dict(
+            self,
+            raise_on_failure=self.raise_on_failure,
+            user_agents=self.user_agents,
+            retry_attempts=self.retry_attempts,
+            timeout=self.timeout,
+        )
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "LinkContentFetcher":
+        """
+        Deserialize this component from a dictionary.
+        """
+        return default_from_dict(cls, data)
+
+    @component.output_types(documents=Optional[Document])
+    def run(self, url: str):
+        """
+        Fetches content from a URL and converts it to a Document objects. If no content is extracted,
+        an empty Document object is returned (if raise_on_failure is False).
```

Comment:
And what happens if raise_on_failure is True? An exception is raised, but what does it mean practically?

Reply:
Well if an exception is raised, the component and/or the pipeline crashes. That's true for all exceptions btw